op-node/rollup/derive/l1_block_info.go

```diff
@@ -143,30 +154,167 @@ func (info *L1BlockInfo) UnmarshalBinary(data []byte) error {
+	if info.BlockHash, err = solabi.ReadHash(r); err != nil {
+		return err
+	}
+	// The "batcherHash" will be correctly parsed as address, since the version 0 and left-padding matches the ABI encoding format.
+	if info.BatcherAddr, err = solabi.ReadAddress(r); err != nil {
+		return err
+	}
+	if info.ValidatorRewardScalar, err = solabi.ReadEthBytes32(r); err != nil {
+		return err
+	}
+	if !solabi.EmptyReader(r) {
+		return errors.New("too many bytes")
+	}
+	return nil
+}
+
+// isEcotoneButNotFirstBlock returns whether the specified block is subject to the Ecotone upgrade,
+// but is not the actiation block itself.
+func isEcotoneButNotFirstBlock(rollupCfg *rollup.Config, l2BlockTime uint64) bool {
+	return rollupCfg.IsEcotone(l2BlockTime) && !rollupCfg.IsEcotoneActivationBlock(l2BlockTime)
+}
+
+// L1BlockInfoFromBytes is the inverse of L1InfoDeposit, to see where the L2 chain is derived from
+func L1BlockInfoFromBytes(rollupCfg *rollup.Config, l2BlockTime uint64, data []byte) (*L1BlockInfo, error) {
 	var info L1BlockInfo
-	err := info.UnmarshalBinary(data)
-	return info, err
+	if isEcotoneButNotFirstBlock(rollupCfg, l2BlockTime) {
+		return &info, info.unmarshalBinaryEcotone(data)
+	}
+	return &info, info.unmarshalBinaryBedrock(data)
 }
 
 // L1InfoDeposit creates a L1 Info deposit transaction based on the L1 block,
 // and the L2 block-height difference with the start of the epoch.
-func L1InfoDeposit(seqNumber uint64, block eth.BlockInfo, sysCfg eth.SystemConfig, regolith bool) (*types.DepositTx, error) {
-	infoDat := L1BlockInfo{
-		Number:                block.NumberU64(),
-		Time:                  block.Time(),
-		BaseFee:               block.BaseFee(),
-		BlockHash:             block.Hash(),
-		SequenceNumber:        seqNumber,
-		BatcherAddr:           sysCfg.BatcherAddr,
-		L1FeeOverhead:         sysCfg.Overhead,
-		L1FeeScalar:           sysCfg.Scalar,
-		ValidatorRewardScalar: sysCfg.ValidatorRewardScalar,
-	}
-	data, err := infoDat.MarshalBinary()
-	if err != nil {
-		return nil, err
+func L1InfoDeposit(rollupCfg *rollup.Config, sysCfg eth.SystemConfig, seqNumber uint64, block eth.BlockInfo, l2BlockTime uint64) (*types.DepositTx, error) {
+	l1BlockInfo := L1BlockInfo{
+		Number:         block.NumberU64(),
+		Time:           block.Time(),
+		BaseFee:        block.BaseFee(),
+		BlockHash:      block.Hash(),
+		SequenceNumber: seqNumber,
+		BatcherAddr:    sysCfg.BatcherAddr,
+	}
+	var data []byte
+	if isEcotoneButNotFirstBlock(rollupCfg, l2BlockTime) {
+		l1BlockInfo.BlobBaseFee = block.BlobBaseFee()
+		if l1BlockInfo.BlobBaseFee == nil {
+			// The L2 spec states to use the MIN_BLOB_GASPRICE from EIP-4844 if not yet active on L1.
+			l1BlockInfo.BlobBaseFee = big.NewInt(1)
+		}
+		blobBaseFeeScalar, baseFeeScalar, err := sysCfg.EcotoneScalars()
+		if err != nil {
+			return nil, err
+		}
+		l1BlockInfo.BlobBaseFeeScalar = blobBaseFeeScalar
+		l1BlockInfo.BaseFeeScalar = baseFeeScalar
+		l1BlockInfo.ValidatorRewardScalar = [32]byte{}
```

Comment:
This means that `ValidatorRewardScalar` is set to 0 after Ecotone?

Reply:
This should be fixed: the value should be injected after ecotone to be correct.